File: designsystem/src/androidMain/kotlin/com/ivangarzab/kluvs/designsystem/components/EmptyState.kt
package com.ivangarzab.kluvs.designsystem.components

import android.annotation.SuppressLint
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.unit.dp
import com.ivangarzab.kluvs.designsystem.components.buttons.OutlinedButton
import com.ivangarzab.kluvs.designsystem.theme.KluvsTheme
import com.ivangarzab.kluvs.designsystem.theme.feature
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.random.Random

// design-system/docs/states.md § Fragmented Hex Grid — keep these in sync with the reference
// asset (design-system/assets/illustration-empty-hexagons.svg) and colors_and_type.css.
// Shared by every EmptyState call site today, on purpose — not yet exposed as a per-call
// parameter. If a screen ever wants its own distinct pattern instead of the incidental
// variation that already comes from differing container sizes, promote this to an optional
// `seed: Int = HEX_SEED` parameter on EmptyState() rather than editing the constant in place.
private const val HEX_SEED = 44
private const val CELL_ACTIVATION_FRACTION = 0.26f
private const val EDGE_KEEP_PROBABILITY = 0.6f
private val HEX_RADIUS = 34.dp
private const val STROKE_WIDTH_DP = 1.75f

// Safe zone as a fraction of the container, matching the reference asset's 360x620 canvas
// (safe rect (40,220)-(320,400)): the illustration must never draw under the centered text.
private const val SAFE_LEFT_FRACTION = 40f / 360f
private const val SAFE_RIGHT_FRACTION = 320f / 360f
private const val SAFE_TOP_FRACTION = 220f / 620f
private const val SAFE_BOTTOM_FRACTION = 400f / 620f

private data class HexEdge(val start: Offset, val end: Offset)

/**
 * Generates the Fragmented Hex Grid pattern for a container of the given size, using a fixed
 * seed so the layout is identical across recompositions and app launches — never regenerated
 * at runtime with a fresh random draw. Cells are activated first, then most of each active
 * cell's edges are kept together; sampling edges independently across the whole grid instead
 * produces disconnected floating lines rather than readable partial-hexagon fragments (see
 * design-system/docs/states.md for the full rationale).
 */
private fun generateHexEdges(widthPx: Float, heightPx: Float, hexRadiusPx: Float): List<HexEdge> {
    val random = Random(HEX_SEED)
    val sqrt3 = sqrt(3f)

    fun hexCenter(row: Int, col: Int): Offset {
        val x = hexRadiusPx * sqrt3 * (col + 0.5f * (row % 2))
        val y = hexRadiusPx * 1.5f * row
        return Offset(x, y)
    }

    fun hexVertices(center: Offset): List<Offset> = (0 until 6).map { i ->
        val angle = Math.toRadians((-90 + 60 * i).toDouble())
        Offset(
            x = center.x + hexRadiusPx * cos(angle).toFloat(),
            y = center.y + hexRadiusPx * sin(angle).toFloat(),
        )
    }

    val safeLeft = widthPx * SAFE_LEFT_FRACTION
    val safeRight = widthPx * SAFE_RIGHT_FRACTION
    val safeTop = heightPx * SAFE_TOP_FRACTION
    val safeBottom = heightPx * SAFE_BOTTOM_FRACTION

    fun touchesSafeZone(p1: Offset, p2: Offset): Boolean =
        listOf(p1, p2).any { it.x in safeLeft..safeRight && it.y in safeTop..safeBottom }

    val rows = (heightPx / (hexRadiusPx * 1.5f)).toInt() + 3
    val cols = (widthPx / (hexRadiusPx * sqrt3)).toInt() + 3

    data class Cell(val center: Offset, val edges: List<Pair<Offset, Offset>>)

    val cells = mutableListOf<Cell>()
    for (r in -1 until rows) {
        for (c in -1 until cols) {
            val center = hexCenter(r, c)
            if (center.x < -hexRadiusPx || center.x > widthPx + hexRadiusPx ||
                center.y < -hexRadiusPx || center.y > heightPx + hexRadiusPx
            ) continue
            val verts = hexVertices(center)
            val edges = (0 until 6).map { i -> verts[i] to verts[(i + 1) % 6] }
            cells.add(Cell(center, edges))
        }
    }

    val eligible = cells.filter { cell ->
        !(cell.center.x > safeLeft - hexRadiusPx && cell.center.x < safeRight + hexRadiusPx &&
            cell.center.y > safeTop - hexRadiusPx && cell.center.y < safeBottom + hexRadiusPx)
    }.shuffled(random)

    val active = eligible.take((eligible.size * CELL_ACTIVATION_FRACTION).toInt())

    val kept = mutableListOf<HexEdge>()
    for (cell in active) {
        for ((p1, p2) in cell.edges) {
            if (touchesSafeZone(p1, p2)) continue
            if (random.nextFloat() < EDGE_KEEP_PROBABILITY) {
                kept.add(HexEdge(p1, p2))
            }
        }
    }
    return kept
}

/**
 * The reusable "nothing here" state — design-system/docs/states.md § Empty State. One
 * illustration (Fragmented Hex Grid) reused everywhere; only heading, body, and the optional
 * [action] slot change per screen. [action] is left null for states with no real next step
 * (e.g. "no search results yet") rather than filled with a button that goes nowhere useful.
 */
@SuppressLint("UnusedBoxWithConstraintsScope")
@Composable
fun EmptyState(
    heading: String,
    body: String,
    modifier: Modifier = Modifier,
    action: @Composable (() -> Unit)? = null,
) {
    val strokeColor = KluvsTheme.colors.disabled
    val density = LocalDensity.current

    // Sizing is entirely the caller's responsibility via [modifier] — fillMaxSize() for a
    // full-screen empty state, a bounded height for a section-level one (e.g. inside a tab
    // that already has content above it).
    BoxWithConstraints(modifier = modifier) {
        val widthPx = with(density) { maxWidth.toPx() }
        val heightPx = with(density) { maxHeight.toPx() }
        val hexRadiusPx = with(density) { HEX_RADIUS.toPx() }
        val strokeWidthPx = with(density) { STROKE_WIDTH_DP.dp.toPx() }
        val edges = remember(widthPx, heightPx) { generateHexEdges(widthPx, heightPx, hexRadiusPx) }

        Canvas(modifier = Modifier.fillMaxSize()) {
            edges.forEach { edge ->
                drawLine(
                    color = strokeColor,
                    start = edge.start,
                    end = edge.end,
                    strokeWidth = strokeWidthPx,
                    cap = StrokeCap.Round,
                )
            }
        }

        Column(
            modifier = Modifier
                .align(Alignment.Center)
                .padding(horizontal = 40.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Text(
                text = heading,
                style = KluvsTheme.typography.title.large.feature(),
                color = KluvsTheme.colors.labelVariant,
                textAlign = TextAlign.Center,
            )
            Spacer(modifier = Modifier.height(10.dp))
            Text(
                text = body,
                style = KluvsTheme.typography.body.medium,
                color = KluvsTheme.colors.contentMuted,
                textAlign = TextAlign.Center,
            )
            if (action != null) {
                Spacer(modifier = Modifier.height(16.dp))
                action()
            }
        }
    }
}

@PreviewLightDark
@Composable
private fun Preview_EmptyState_NoAction() = KluvsTheme {
    EmptyState(
        modifier = Modifier.fillMaxSize().background(KluvsTheme.colors.background),
        heading = "Nothing shelved yet.",
        body = "Search for a book and add it to Want to Read, Read, or Not Finished.",
    )
}

@PreviewLightDark
@Composable
private fun Preview_EmptyState_WithAction() = KluvsTheme {
    EmptyState(
        modifier = Modifier.fillMaxSize().background(KluvsTheme.colors.background),
        heading = "No clubs yet.",
        body = "Join a club or start your own to get things going.",
        action = { OutlinedButton(text = "Join with a code", onClick = {}) },
    )
}

File: iosApp/DesignSystem/Sources/DesignSystem/Components/States/EmptyState.swift
import SwiftUI

/// Deterministic seeded PRNG (SplitMix64) — Swift's `SystemRandomNumberGenerator` cannot be
/// seeded, so a fixed-seed pattern (see `EmptyState` below) needs its own generator. Does not
/// need to produce bit-identical output to Android's `kotlin.random.Random(seed)` — each
/// platform only needs to be deterministic against itself, not against the other platform's
/// PRNG algorithm (see design-system/docs/states.md).
private struct SeededGenerator: RandomNumberGenerator {
    private var state: UInt64

    init(seed: Int) {
        state = UInt64(bitPattern: Int64(seed)) &+ 0x9E3779B97F4A7C15
    }

    mutating func next() -> UInt64 {
        state &+= 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return z ^ (z >> 31)
    }
}

private struct HexEdge {
    let start: CGPoint
    let end: CGPoint
}

// design-system/docs/states.md § Fragmented Hex Grid — keep these in sync with Android's
// EmptyState.kt and the reference asset (design-system/assets/illustration-empty-hexagons.svg).
// Shared by every EmptyState call site today, on purpose — not yet exposed as a per-call
// parameter. See Android's equivalent constant for the promotion note if that ever changes.
private let hexSeed = 44
private let cellActivationFraction = 0.26
private let edgeKeepProbability = 0.6
private let hexRadius: CGFloat = 34
private let strokeWidth: CGFloat = 1.75

// Safe zone as a fraction of the container, matching the reference asset's 360x620 canvas
// (safe rect (40,220)-(320,400)): the illustration must never draw under the centered text.
private let safeLeftFraction: CGFloat = 40 / 360
private let safeRightFraction: CGFloat = 320 / 360
private let safeTopFraction: CGFloat = 220 / 620
private let safeBottomFraction: CGFloat = 400 / 620

/// Generates the Fragmented Hex Grid pattern for a container of the given size, using a fixed
/// seed so the layout is identical across renders and app launches — never regenerated at
/// runtime with a fresh random draw. Cells are activated first, then most of each active cell's
/// edges are kept together; sampling edges independently across the whole grid instead produces
/// disconnected floating lines rather than readable partial-hexagon fragments (see
/// design-system/docs/states.md for the full rationale). Mirrors Android's `generateHexEdges`.
private func generateHexEdges(width: CGFloat, height: CGFloat) -> [HexEdge] {
    var rng = SeededGenerator(seed: hexSeed)
    let sqrt3 = CGFloat(3).squareRoot()

    func hexCenter(row: Int, col: Int) -> CGPoint {
        let x = hexRadius * sqrt3 * (CGFloat(col) + 0.5 * CGFloat(row % 2))
        let y = hexRadius * 1.5 * CGFloat(row)
        return CGPoint(x: x, y: y)
    }

    func hexVertices(center: CGPoint) -> [CGPoint] {
        (0..<6).map { i in
            let angle = (-90.0 + 60.0 * Double(i)) * .pi / 180.0
            return CGPoint(
                x: center.x + hexRadius * CGFloat(cos(angle)),
                y: center.y + hexRadius * CGFloat(sin(angle))
            )
        }
    }

    let safeLeft = width * safeLeftFraction
    let safeRight = width * safeRightFraction
    let safeTop = height * safeTopFraction
    let safeBottom = height * safeBottomFraction

    func touchesSafeZone(_ p1: CGPoint, _ p2: CGPoint) -> Bool {
        [p1, p2].contains { p in
            p.x >= safeLeft && p.x <= safeRight && p.y >= safeTop && p.y <= safeBottom
        }
    }

    let rows = Int(height / (hexRadius * 1.5)) + 3
    let cols = Int(width / (hexRadius * sqrt3)) + 3

    struct Cell {
        let center: CGPoint
        let edges: [(CGPoint, CGPoint)]
    }

    var cells: [Cell] = []
    for r in -1..<rows {
        for c in -1..<cols {
            let center = hexCenter(row: r, col: c)
            if center.x < -hexRadius || center.x > width + hexRadius ||
                center.y < -hexRadius || center.y > height + hexRadius {
                continue
            }
            let verts = hexVertices(center: center)
            let edges = (0..<6).map { i in (verts[i], verts[(i + 1) % 6]) }
            cells.append(Cell(center: center, edges: edges))
        }
    }

    let eligible = cells.filter { cell in
        !(cell.center.x > safeLeft - hexRadius && cell.center.x < safeRight + hexRadius &&
            cell.center.y > safeTop - hexRadius && cell.center.y < safeBottom + hexRadius)
    }.shuffled(using: &rng)

    let active = eligible.prefix(Int(Double(eligible.count) * cellActivationFraction))

    var kept: [HexEdge] = []
    for cell in active {
        for (p1, p2) in cell.edges {
            if touchesSafeZone(p1, p2) { continue }
            if Double.random(in: 0..<1, using: &rng) < edgeKeepProbability {
                kept.append(HexEdge(start: p1, end: p2))
            }
        }
    }
    return kept
}

/// The reusable "nothing here" / "couldn't load this" state shell — design-system/docs/states.md
/// § Empty State. One illustration (Fragmented Hex Grid) reused everywhere; only heading, body,
/// `lineColor`, and the optional `action` slot change per screen. Omit `action` for states with
/// no real next step (e.g. "no search results yet") rather than filling it with a button that
/// goes nowhere useful. `lineColor` defaults to the quiet empty-state tone; `ErrorView` overrides
/// it to `danger` red so the same illustration reads as a failure, not absence. Mirrors Android's
/// `EmptyState` composable.
public struct EmptyState<Action: View>: View {
    private let heading: String
    private let bodyText: String
    private let lineColor: Color
    private let action: Action

    public init(
        heading: String,
        body: String,
        lineColor: Color = KluvsTheme.colors.disabled,
        @ViewBuilder action: () -> Action = { EmptyView() }
    ) {
        self.heading = heading
        self.bodyText = body
        self.lineColor = lineColor
        self.action = action()
    }

    public var body: some View {
        GeometryReader { geometry in
            ZStack {
                Canvas { context, size in
                    let edges = generateHexEdges(width: size.width, height: size.height)
                    var path = Path()
                    for edge in edges {
                        path.move(to: edge.start)
                        path.addLine(to: edge.end)
                    }
                    context.stroke(
                        path,
                        with: .color(lineColor),
                        style: StrokeStyle(lineWidth: strokeWidth, lineCap: .round)
                    )
                }

                VStack(spacing: 10) {
                    Text(heading)
                        .kluvsStyle(KluvsTheme.typography.title.large, feature: true)
                        .foregroundColor(KluvsTheme.colors.labelVariant)
                        .multilineTextAlignment(.center)
                    Text(bodyText)
                        .kluvsStyle(KluvsTheme.typography.body.medium)
                        .foregroundColor(KluvsTheme.colors.contentMuted)
                        .multilineTextAlignment(.center)
                    action
                        .padding(.top, 6)
                }
                .padding(.horizontal, 40)
                .frame(width: geometry.size.width)
            }
        }
    }
}

#Preview("EmptyState — no action") {
    EmptyState(
        heading: "Nothing shelved yet.",
        body: "Search for a book and add it to Want to Read, Read, or Not Finished."
    )
    .frame(width: 360, height: 620)
    .background(KluvsTheme.colors.background)
}

#Preview("EmptyState — with action") {
    EmptyState(
        heading: "No clubs yet.",
        body: "Join a club or start your own to get things going."
    ) {
        OutlinedButton(text: "Join with a code", action: {})
    }
    .frame(width: 360, height: 620)
    .background(KluvsTheme.colors.background)
}
